### PR TITLE
carbonara: use numpy native types for computing

### PR DESCRIPTION
**Pull request by jd**
This replaces various weak data type (e.g. integers for sampling are replaced
by numpy.timedelta64) by native numpy data types.

Pandas data types are replaced by numpy data types where possible.

Change-Id: I923d5d4fedf571deb4dac74a72cdd539c51cc5a9

### CODE REVIEW

File: gnocchi/carbonara.py [hunk 1]
```diff
@@ -73,19 +76,23 @@ def __init__(self):
...
 
 
+def datetime64_to_epoch(dt):
+    return (dt - UNIX_UNIVERSAL_START64) / ONE_SECOND
+
+
 def round_timestamp(ts, freq):
-    return pandas.Timestamp(
-        (pandas.Timestamp(ts).value // freq) * freq)
+    return UNIX_UNIVERSAL_START64 + numpy.floor(
```
Comment (chungg) [inline gnocchi/carbonara.py]: UNIX_UNIVERSAL_START64 doesn't seem to do anything... it resolves to 0

```
>>> ts = numpy.array(rng[:5], dtype=numpy.datetime64)
>>> ts
array(['2012-01-01T00:00:00.000000000', '2012-01-01T00:00:01.000000000',
       '2012-01-01T00:00:02.000000000', '2012-01-01T00:00:03.000000000',
       '2012-01-01T00:00:04.000000000'], dtype='datetime64[ns]')
>>> ts - np.datetime64("1970", 'ns')
array([1325376000000000000, 1325376001000000000, 1325376002000000000,
       1325376003000000000, 1325376004000000000], dtype='timedelta64[ns]')
>>> ts.astype(int)
array([1325376000000000000, 1325376001000000000, 1325376002000000000,
       1325376003000000000, 1325376004000000000])
>>> np.datetime64("1970", 'ns').astype(int)
0
```

File: gnocchi/utils.py [hunk 2]
```diff
@@ -140,12 +140,21 @@ def to_timespan(value):
...
         raise ValueError("Timespan must be positive")
-    return datetime.timedelta(seconds=seconds)
+    return seconds
+
+
+_ONE_SECOND = numpy.timedelta64(1, 's')
+
+
+def timespan_total_seconds(td):
+    return td / _ONE_SECOND
```
Comment (chungg) [inline gnocchi/utils.py]: just use astype('timedelta64[s]')

i'm assuming you want timedelta64[s]

File: gnocchi/carbonara.py (hunk 1 above)
Comment (jd) [inline gnocchi/carbonara.py]: This is used for type conversion. The result on right side of `+` is a `numpy.timedelta64`, so in order to convert it, we need to add it to what is the 0 time indeed.

File: gnocchi/utils.py (hunk 2 above)
Comment (jd) [inline gnocchi/utils.py]: >>> numpy.timedelta64(100, 'ns').astype('timedelta64[s]').astype(int)
0

**Comment by jd** [comment]
We don't use of Carbonara and what uses Carbonara. I don't understand what you mean?

**Comment by chungg** [comment]
ie. this. if you look at archive_policy.py, why does ArchivePolicy need to be know anything about numpy? there's a bunch of places ArchivePolicy is either given numpy OR a datetime.timedelta that is just casting numpy.timedelta

**Comment by jd** [comment]
Because Carbonara excepts Numpy types, and all the storage layer is built around Carbonara. There's no need to have intermediary float/int type whereas the type are actually `datetime.timedelta` (which don't have nanoseconds precision – so let's just use `numpy.timedelta64`.

We can merge https://github.com/gnocchixyz/gnocchi/pull/209 first if you feel more comfortable with asserting we depend on Carbonara.

**Comment by chungg** [comment]
i'll keep looking but this unrelated to #209. regardless of #209, everything outside of carbonara is either bytes or string representation of epoch. i would think if a carbonara timeseries wants numpy, it should convert/serialise as it needs on creation. it just looks super strange that archivepolicy has numpy when it's just a number.

**Comment by jd** [comment]
> i'll keep looking but this unrelated to #209. regardless of #209, everything outside of carbonara is either bytes or string representation of epoch.

This is not true with this branch/patch and it's exactly why this patch is *awesome*.

> i would think if a carbonara timeseries wants numpy, it should convert/serialise as it needs on creation. it just looks super strange that archivepolicy has numpy when it's just a number.

It's not a *number*. It's a *timedelta with a nanoseconds precision*. Python does not have it. Numpy has it. So… I use Numpy :)

File: gnocchi/utils.py (hunk 2 above)
Comment (chungg) [inline gnocchi/utils.py]: i don't understand. the functions returns seconds. what do you expect 100ns to resolve to in seconds?

**Comment by chungg** [comment]
> It's not a number. It's a timedelta with a nanoseconds precision. Python does not have it. Numpy has it. So… I use Numpy :)

you sure? i'm pretty sure it's just a string epoch in indexer and in object name/key.

File: gnocchi/utils.py (hunk 2 above)
Comment (jd) [inline gnocchi/utils.py]: It returns an integer of the total in seconds.
`numpy.timedelta64(100, 'ns')` should be converted to `0.00000001`.

**Comment by jd** [comment]
> you sure? i'm pretty sure it's just a string epoch in indexer and in object name/key.

In indexer, native type of MySQL/PostgreSQL are used and then `datetime.timedelta` are returned. We don't have (nor need I'd say) nanoseconds precision so that's good enough. We anyway don't use epoch from the indexer in any storage place, so it's really two different things.

In object name, it's `SplitKey`s that are passed around. To construct the final object name in the driver, they are converted to a string with `str` when, as late as possible, an object name is computed. As soon as a key is read (see `list_split_key`), it's converted to a SplitKey (with a `numpy.datetime64` in it). This work has been done in the first patch of the serie.

So there's nowhere a string/float carried anymore. That's one of the point of this branch, to stop mixing string, float, and datetimes, and have a **correct** type: a timedelta/datetime with nanosecond precision, everywhere. :)

File: gnocchi/utils.py (hunk 2 above)
Comment (chungg) [inline gnocchi/utils.py]: right... but what scenario do you need 0.0000001? you're supporting a scenario which the storage can't support... unless your plan is to support nanosecond aggregates?

Comment (jd) [inline gnocchi/utils.py]: We always supported nanoseconds aggregate. :)

**Comment by sileht** [comment]
I also prefer the numpy everywhere, I have never liked the API-string->datetime.datetime->pandas.Timestamp->string/int-for-split-key->numpy.timestamp-for-serialization and the reverse.

pandas is cool when you have ton of mixed unsorted data. But for simple thing like us, ordered two columns table, timestamp/float, it's overkill and pandas does not always choice the good algorithm because it always assume data are mixed and unsorted. We already improve that by removing it from aggregation computing and serialization/unserialization.

This change just goes one step further.

**Comment by chungg** [comment]
ok, im starting to see the number of places we need to do archivepolicy conversion.

that said, i've found another concern. unserialisation speed is consistently, noticeably slower in the benchmarks. testing the aggregated series, it's about ~30% slower but the bounded series benchmark is ~400% slower. i don't know what's happening here but it's visible in the gate and in my local environment.

**Comment by jd** [comment]
> that said, i've found another concern. unserialisation speed is consistently, noticeably slower in the benchmarks. testing the aggregated series, it's about ~30% slower but the bounded series benchmark is ~400% slower. i don't know what's happening here but it's visible in the gate and in my local environment.

Good point. So be aware that most benchmarks are not reliable as they use `random()` and that do changes a lot the performance of the different computing and compression behind. So changes of 20-30% in various test run is expected.

The goal of the benchmark is more to notice drawdown of at 2x and more.

That being said, I found one benchmark that is considerably slow, and it's indeed the `BoundTimeSerie` serialization. After digging a bit, I did not find why it was slower, but I found how to make it as fast as previously:

I've modified this patch to add the following:
```python
        # FIXME(jd) Return the result as a pandas.Timestamp object as Pandas is
        # faster at indexing pandas.Timestamp than numpy.timedelta64 objects
        # for whatever reason
        return pandas.Timestamp(
            rounded - (self.block_size * self.back_window)
        )
```

So we clearly need to continue ditching Pandas…

**Comment by jd** [comment]
Pull-request updated, HEAD is now 1931cf240aa5ce8d19293691e8804b1d28d691a3

**Comment by chungg** [comment]
> Good point. So be aware that most benchmarks are not reliable as they use random() and that do changes a lot the performance of the different computing and compression behind. So changes of 20-30% in various test run is expected.

i don't think that's true. i usually see maybe 5-10% deviation on the same code. especially when looking at stuff like 'all zeros'. i would say 20-30% deviation is notable change in performance.

**Comment by chungg** [comment]
the 4x slow down it better now though :)

**Comment by jd** [comment]
> i don't think that's true. i usually see maybe 5-10% deviation on the same code.

I really saw weird jumps in perf, mainly because of lz4 I presume, when things are random, much more than 5-10%.

> especially when looking at stuff like 'all zeros'. i would say 20-30% deviation is notable change in performance.

For all 0 / all 1 (everything that is not random) that's true, yep (that's what I wrote).

File: gnocchi/tests/functional/gabbits/archive.yaml [hunk 3]
```diff
@@ -561,8 +561,8 @@ tests:
       status: 201
       response_json_paths:
           $.definition[0].points: 1000
-          $.definition[0].granularity: "0:00:04"
-          $.definition[0].timespan: "1:06:40"
+          $.definition[0].granularity: "0:00:03.600000"
+          $.definition[0].timespan: "1:00:00"
```
Comment (chungg) [inline gnocchi/tests/functional/gabbits/archive.yaml]: i guess this was broken before?

File: gnocchi/tests/test_carbonara.py [hunk 4]
```diff
@@ -19,13 +19,17 @@
...
 import iso8601
-import pandas
+import numpy
 import six
 
 from gnocchi import carbonara
 from gnocchi.tests import base
 
 
+def datetime64(*args):
```
Comment (chungg) [inline gnocchi/tests/test_carbonara.py]: is this change necessary?

Comment (chungg) [inline gnocchi/tests/test_carbonara.py]: for the non-assert stuff i mean... i guess it doesn't matter... it's a sh*tload of code regardless

File: gnocchi/tests/test_storage.py [hunk 5]
```diff
@@ -354,34 +398,58 @@ def test_rewrite_measures(self):
         self.trigger_processing()
 
         self.assertEqual({
-            carbonara.SplitKey(1452384000.0, 60),
-            carbonara.SplitKey(1451736000.0, 60),
-            carbonara.SplitKey(1451520000.0, 60),
-            carbonara.SplitKey(1451952000.0, 60),
-        }, self.storage._list_split_keys_for_metric(self.metric, "mean", 60.0))
+            carbonara.SplitKey(numpy.datetime64('2016-01-10T00:00:00'),
```
Comment (chungg) [inline gnocchi/tests/test_storage.py]: maybe keep it as epoch so we can match it with the rest of the test.

File: gnocchi/tests/test_storage.py [hunk 6]
```diff
@@ -404,33 +472,51 @@ def test_rewrite_measures_oldest_mutable_timestamp_eq_next_key(self):
         self.trigger_processing()
 
         self.assertEqual({
-            carbonara.SplitKey(1451520000.0, 60),
-            carbonara.SplitKey(1451736000.0, 60),
-            carbonara.SplitKey(1451952000.0, 60),
-        }, self.storage._list_split_keys_for_metric(self.metric, "mean", 60.0))
+            carbonara.SplitKey(numpy.datetime64('2015-12-31T00:00:00'),
```
Comment (chungg) [inline gnocchi/tests/test_storage.py]: same thing

File: gnocchi/tests/test_storage.py [hunk 7]
```diff
@@ -445,33 +531,56 @@ def test_rewrite_measures_oldest_mutable_timestamp_eq_next_key(self):
         self.trigger_processing()
 
         self.assertEqual({
-            carbonara.SplitKey(1452384000.0, 60),
-            carbonara.SplitKey(1451736000.0, 60),
-            carbonara.SplitKey(1451520000.0, 60),
-            carbonara.SplitKey(1451952000.0, 60),
-        }, self.storage._list_split_keys_for_metric(self.metric, "mean", 60.0))
+            carbonara.SplitKey(numpy.datetime64('2016-01-10T00:00:00'),
```
Comment (chungg) [inline gnocchi/tests/test_storage.py]: same

File: gnocchi/json.py [hunk 8]
```diff
@@ -33,6 +34,15 @@ def to_primitive(obj):
     if isinstance(obj, numpy.datetime64):
         # Do not include nanoseconds if null
         return str(obj).rpartition(".000000000")[0] + "+00:00"
+    if isinstance(obj, numpy.timedelta64):
+        return obj / numpy.timedelta64(1, 's')
+    if isinstance(obj, pandas.Timedelta):
+        # >>> pandas.Timedelta("1 minute").total_seconds()
+        # 60.00000000000001
+        # Never forget that bro.
+        return to_primitive(obj.to_timedelta64())
```
Comment (jd) [inline gnocchi/json.py]: Changed, thanks!

File: gnocchi/tests/functional/gabbits/archive.yaml (hunk 3 above)
Comment (jd) [inline gnocchi/tests/functional/gabbits/archive.yaml]: Yes. I read that twice but we always supported nanoseconds aggregation so if you request 1 hours and 1k points this is what will happen. I guess it was not clear when we wrote the test that nanoseconds aggregation was expected to work and no rounding was done.

The underlying aggregation done was already 3.6s so that does not change. We just fix the returned result on this API call.

File: gnocchi/tests/test_carbonara.py (hunk 4 above)
Comment (jd) [inline gnocchi/tests/test_carbonara.py]: Yup. It's just search-and-replace everywhere anyway.

File: gnocchi/tests/test_storage.py (hunk 5 above)
Comment (jd) [inline gnocchi/tests/test_storage.py]: Changed those back to epoch.

File: gnocchi/tests/test_storage.py (hunk 6 above)
Comment (jd) [inline gnocchi/tests/test_storage.py]: Fixed

File: gnocchi/tests/test_storage.py (hunk 7 above)
Comment (jd) [inline gnocchi/tests/test_storage.py]: Fixed

**Comment by jd** [comment]
@chungg I digged a bit in the benchmark stuff but did not find anything obvious. The code has barely been edited for serialization here but it seems slower. The slowest operation seems to be `         serial[locs] = values` in serialization, but it's all numpy, before this patch, and after this patch. So maybe a few others operations are slowed down, but it's barely possible to tell.

I tried profiling yesterday but it's mainly optimized (and therefore invisible) numpy calls so it's barely possible to get anything useful.

I suspect it's another Pandas side effect like `BoundTimeSeries`. If you're not comfortable merging this as is, I'm OK to wait until more Pandas is removed and performance are back to (and I hope better) than previously. Though it means it'll make that PR bigger. ;)

**Comment by jd** [comment]
Pull-request updated, HEAD is now 754ce1f1dee5c4cdf867fca7c1e17f89ce61446c

**Comment by jd** [comment]
Rebase because of rate: agg method add

File: gnocchi/carbonara.py [hunk 9]
```diff
@@ -151,7 +157,7 @@ def quantile(self, q):
                                          numpy.percentile,
                                          q=q,
                                      ),
-                                     out_dtype='float64',
+                                     out_dtype='float6y4',
```
Comment (chungg) [inline gnocchi/carbonara.py]: guessing this is stray character

**Comment by jd** [comment]
Fixed typo spotted by gordc

**Comment by jd** [comment]
Fixed tests after rate: patch

**Comment by jd** [comment]
Pull-request updated, HEAD is now e7d080ee71220629ff1558711d9bb0971d74ad80

**Comment by jd** [comment]
It seems I cant get anything working today
